File: solutions/elixir/raindrops/1/lib/raindrops.ex
defmodule Raindrops do
  @doc """
  Returns a string based on raindrop factors.

  - If the number contains 3 as a prime factor, output 'Pling'.
  - If the number contains 5 as a prime factor, output 'Plang'.
  - If the number contains 7 as a prime factor, output 'Plong'.
  - If the number does not contain 3, 5, or 7 as a prime factor,
    just pass the number's digits straight through.
  """
  @spec convert(pos_integer) :: String.t()
  def convert(number) do
    factor_3(number) <> factor_5(number) <> factor_7(number) <> factor_none(number)
  end

  def factor_3(number) when rem(number, 3) == 0, do: "Pling"
  def factor_3(_number), do: ""

  def factor_5(number) when rem(number, 5) == 0, do: "Plang"
  def factor_5(_number), do: ""

  def factor_7(number) when rem(number, 7) == 0, do: "Plong"
  def factor_7(_number), do: ""

  def factor_none(number)
      when rem(number, 3) == 0 or
             rem(number, 5) == 0 or
             rem(number, 7) == 0,
      do: ""

  def factor_none(number), do: to_string(number)
end

File: solutions/lua/raindrops/2/raindrops.lua
local raindrops = {
  [3] = 'Pling',
  [5] = 'Plang',
  [7] = 'Plong',
}
local keys = {}
for key in pairs(raindrops) do
  table.insert(keys, key)
end
table.sort(keys)

return function(n)
  local result = ''
  for _, key in ipairs(keys) do
    if n % key == 0 then
      result = result..raindrops[key]
    end
  end
  if result == '' then 
    return tostring(n)
  end
  return result
end